### DISCUSSION TITLE
Qudi-modules with R&S®SMBV100A

### DISCUSSION THREAD

**Discussion by LiamHanlon** [Ulm-IQO/qudi-core#72] [Q&A] [answered]
Has anyone made the migration to qudi-core/modules using a R&S®SMBV100A? 

The ODMR scan worked fine in the old version, but with the new version, I get a lot of errors that seem to come from the hardware file mw_source_smbv.py file itself. 

First the system could not recognize the model which we think was due to a typo: 
line 83: self.model was changed to self._model

Then it could not find a scan mode which we added ourselves: 
line 93: scan_modes=(SamplingOutputMode.EQUIDISTANT_SWEEP)
changed to add the other mode: scan_modes=(SamplingOutputMode.EQUIDISTANT_SWEEP, SamplingOutputMode.JUMP_LIST) 

The ODMR now seems to work fine in cw mode but I get an error: 'MicrowaveSmbv' object has no attribute '_write_list' line 245.

I have tried a few changes but nothing seems to work. I am hesitant to make so many changes to the source code and making changes to the config file didn't seem to help.

Does anyone else work with this system and has any advice?

**Comment by NicolasStaudenmaier** [comment]
Hi @LiamHanlon, thank you for your questions. Indeed, it seems there are some typos in this hardware file. You are correct that in **line 83** it must read `self._model`. Second, though the mw generator may be capable of using the list mode, only the `EQUIDISTANT_SWEEP` option is implemented. Accordingly, in **line 245** it should be `self._write_sweep()`. 
As a note, please don't add `SamplingOutputMode.JUMP_LIST` in the scan_modes (line 93) unless the full toolchain is implemented. 
To properly run ODMR in the EQUIDISTANT_SWEEP mode, please set the ConfigOption for the ODMR logic. Your configuration entry will look similar to this (note the last or last two lines): 
```
odmr_logic:
    module.Class: 'odmr_logic.OdmrLogic'
    connect:
        microwave: <microwave_name>
        data_scanner: <data_scanner_name>
    options:
        default_scan_mode: 'EQUIDISTANT_SWEEP'
```

Thank you for your help with improving our modules. Any external feedback is much appreciated. 
This hardware device is rarely used in our institute. I also can't test these fixes with the hardware itself at the moment. I hope it works for you. If you run into further problems, just let us know.

**Reply by LiamHanlon** [reply]
Hi @NicolasStaudenmaier , thanks for the response. 

I made the changes you suggested and still got a runtime error: 
RuntimeError: Can not reset frequency scan. CW microwave output active.
Traceback (most recent call last) ...

Turns out there is another small code error on line 312, it reads: 

if self._in_cw_mode: ...
but it should read: 
if not self._in_cw_mode: ...

It seems to work fine now. Thanks for the help

**Reply by NicolasStaudenmaier** [reply]
Thanks for testing and the feedback. 
It is another typo. The True/False logic is correct here but it is misssing (). So it reads `if self._in_cw_mode():`. You find the same typo in line 291 where it should be `if not self._in_cw_mode():`. 
Can you please confirm again that it runs with this change? We will then merge the changes mentioned here in the qudi-iqo-modules repo.

**Reply by LiamHanlon** [reply]
Hi @NicolasStaudenmaier, 

the changes work and the ODMR runs fine. Thanks for the help

**Comment by NicolasStaudenmaier** [comment, marked as answer]
The bugs have been resolved with [PR #107](https://github.com/Ulm-IQO/qudi-iqo-modules/pull/107) in the qudi-iqo-modules repository.
